### PR TITLE
6X: Backport FDW features to support Greenplum FDW

### PR DESCRIPTION
**Pull request by huiliang-liu**
This PR contains two features needed by Greenplum FDW.

1, Support creating a Gang with QEs more than the segment count 

Scenarios like FDW could have more QEs than the segment count to query
the foreign clusters.

2, Support num_segments option for foreign servers

Support specifying the num_segments option for foreign servers,
Greenplum will create that amount of QEs if the `mpp_execute`
option's value is 'all segments'.

```
CREATE SERVER s1 FOREIGN DATA WRAPPER dummy OPTIONS (num_segments '42');
```

### CODE REVIEW

File: src/backend/foreign/foreign.c [hunk 1]
```diff
@@ -69,7 +74,33 @@ SeparateOutMppExecute(List **options)
...
 	}
 
-	return exec_location;
+
+	prev = NULL;
+	foreach(lc, *options)
+	{
+		DefElem    *def = (DefElem *) lfirst(lc);
+
+		if (strcmp(def->defname, "segment_number") == 0)
```
Comment (bradfordb-vmware) [inline src/backend/foreign/foreign.c:83]: If I understand this correctly, we are adding an option called `segment_number` to the `CREATE FOREIGN DATA WRAPPER` statement but it's only use (currently) is for the gp2gp FDW? I am not sure this make sense for two reasons:

1. Not all FDWs will be accessing data on systems that have a concept of number of segments (e.g., PXF FDW for HDFS)
2. The gp2gp FDW might want to access different GPDB clusters, each with a different number of segments.

Is it possible for this to be an option to `CREATE SERVER`?

File: src/backend/libpq/be-gssapi-common.c [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	 * before calling this function anyway.
+	 */
+	ereport(severity,
+			(errmsg_internal("%s", errmsg),
+			 errdetail_internal("%s: %s", msg_major, msg_minor)));
+}
+
+
+void
+pg_store_proxy_credential(gss_cred_id_t cred)
```
[the remark this answers is not in the document]
Comment (huiliang-liu) [reply src/backend/libpq/be-gssapi-common.c:82]: This PR comes from the master PR #12372. It is the kernel part of that PR.
As I known, the KDC server doesn't need to modify for delegation by default. The gss api `gss_init_sec_context `'s `GSS_C_DELEG_FLAG ` permits delegation of the initiator's credentials

[the remark this answers is not in the document]
Comment (huiliang-liu) [reply src/backend/libpq/be-gssapi-common.c:82]: I will prepare a documentation about the proxy. Thank you for your suggestion.

File: src/backend/foreign/foreign.c (hunk 1 above)
Comment (huiliang-liu) [reply src/backend/foreign/foreign.c:83]: That is a good question.
Yes, the `segment_number` option can be used in `CREATE FOREIGN DATA WRAPPER` even for file or pxf or other FDW. It is not good. 
Actually, `segment_number` is an option to `CREATE SERVER` and `CREATE FOREIGN TABLE` now. I will try to limit it only for the two statements except `CREATE FOREIGN DATA WRAPPER`

Comment (huiliang-liu) [reply src/backend/foreign/foreign.c:89]: fixed

File: src/backend/foreign/foreign.c [hunk 3]
```diff
@@ -72,6 +73,41 @@ SeparateOutMppExecute(List **options)
 	return exec_location;
 }
 
+/* Get and separate out the segment_number option */
+int32
+SeparateOutSegmentNumber(List **options)
```
Comment (adam8157) [inline src/backend/foreign/foreign.c:78]: IIRC, we had an implementation that getting this information dynamically, via a query.

Is it? Can we do that? Specifying the segment number manually could be a pain if the customer set it wrongly or the remote cluster got expanded/shrank.

File: src/backend/foreign/foreign.c [hunk 4]
```diff
@@ -294,12 +334,15 @@ GetForeignTable(Oid relid)
...
 
+	ForeignServer *server = GetForeignServer(ft->serverid);
 	ft->exec_location = SeparateOutMppExecute(&ft->options);
 	if (ft->exec_location == FTEXECLOCATION_NOT_DEFINED)
-	{
-		ForeignServer *server = GetForeignServer(ft->serverid);
 		ft->exec_location = server->exec_location;
-	}
+
+	ft->segment_number = SeparateOutSegmentNumber(&ft->options);
```
Comment (adam8157) [inline src/backend/foreign/foreign.c:342]: -1 to this block. If we decide to have a "segment_number" option, it feels like it's only a server option.

File: src/backend/utils/cache/relcache.c [hunk 5]
```diff
@@ -1012,6 +1012,7 @@ RelationBuildDesc(Oid targetRelId, bool insertIt)
      */
     if ((relation->rd_rel->relkind == RELKIND_RELATION &&
 		 !IsSystemRelation(relation)) ||
+		(relation->rd_rel->relkind == RELKIND_FOREIGN_TABLE) ||
```
Comment (adam8157) [inline src/backend/utils/cache/relcache.c:1015]: What is this for?

File: src/backend/foreign/foreign.c (hunk 3 above)
Comment (adam8157) [reply src/backend/foreign/foreign.c:78]: probably I was wrong.

File: src/backend/utils/cache/relcache.c (hunk 5 above)
Comment (huiliang-liu) [reply src/backend/utils/cache/relcache.c:1015]: I have added it(refer to master) when I fix a crash issue in test cases. but I found all the cases pass without this change based on the latest code.
So I will rollback it.

File: src/backend/foreign/foreign.c (hunk 4 above)
Comment (huiliang-liu) [reply src/backend/foreign/foreign.c:342]: This part is just following the master's code. The original design is that `segment_number ` can be used for fdw, server and foreign table. It doesn't make sense for fdw level, like Bradford mentioned, and I have fixed it.
I think the option for foreign table level is needed. it can override the server's configuration for special table which may has diff segment number from server

File: contrib/postgres_fdw/postgres_fdw.c [hunk 6]
```diff
@@ -915,14 +963,84 @@ postgresBeginForeignScan(ForeignScanState *node, int eflags)
...
+	fsstate->is_gp_retrieve =
+		(table->exec_location == FTEXECLOCATION_ALL_SEGMENTS &&
+		 Gp_role == GP_ROLE_EXECUTE);
+
+	fsstate->is_gp_parallel_retrieve_cursor =
+		(table->exec_location == FTEXECLOCATION_ALL_SEGMENTS &&
+		 Gp_role == GP_ROLE_DISPATCH);
+	if (fsstate->is_gp_parallel_retrieve_cursor)
+	{
+		int seg_number = get_remote_segment_number(server, user);
```
Comment (zxuejing) [inline contrib/postgres_fdw/postgres_fdw.c:975]: get_remote_segment_number gets segmentnums from gp_segment_configuration, but segmentnums can be different among tables, 
such as:
create table t3(tc1 int);
create extension gp_debug_numsegments;
select gp_debug_set_create_table_default_numsegments(2);
create table t2(tc1 int);
select * from gp_distribution_policy ;
the table t2 is created on 2 segments,however, t2 is created on 3 segments
we can get real  segment num from gp_distribution_policy;

File: src/test/regress/sql/gp_foreign_data.sql [hunk 7]
```diff
@@ -24,3 +24,12 @@ CREATE FOREIGN TABLE ft3 (
 CREATE FOREIGN TABLE ft4 (
 	c1 int
 ) SERVER s0 OPTIONS (delimiter ',', mpp_execute 'all segments');
+
+-- CREATE FOREIGN SERVER
+CREATE SERVER s1 FOREIGN DATA WRAPPER dummy OPTIONS (segment_number '3');
+
+-- CREATE FOREIGN TABLE
+CREATE FOREIGN TABLE ft5 (
```
[the remark this answers is not in the document]
Comment (huiliang-liu) [reply src/test/regress/sql/gp_foreign_data.sql:32]: Yes, the foreign table's segment_number will override and may be bigger than the value of server

File: src/include/foreign/foreign.h [hunk 8]
```diff
@@ -68,10 +69,12 @@ typedef struct ForeignTable
 	Oid			serverid;		/* server Oid */
 	List	   *options;		/* ftoptions as DefElem list */
 	char		exec_location;  /* execute on MASTER, ANY or ALL SEGMENTS, Greenplum MPP specific */
+	int32		segment_number;	/* segment number of the foreign cluster */
```
[the remark this answers is not in the document]
Comment (huiliang-liu) [reply src/include/foreign/foreign.h:72]: because `segment_number` as well as `exec_location` will be used in planner, it is better to extact and  save the value in `ForeignTable ` .

File: src/backend/optimizer/util/pathnode.c [hunk 9]
```diff
@@ -3065,7 +3065,16 @@ create_foreignscan_path(PlannerInfo *root, RelOptInfo *rel,
 			CdbPathLocus_MakeGeneral(&(pathnode->path.locus), getgpsegmentCount());
 			break;
 		case FTEXECLOCATION_ALL_SEGMENTS:
-			CdbPathLocus_MakeStrewn(&(pathnode->path.locus), getgpsegmentCount());
+			if (rel->ftEntry && rel->ftEntry->segment_number > 0)
+			{
+				CdbPathLocus_MakeStrewn(&(pathnode->path.locus), rel->ftEntry->segment_number);
+				pathnode->path.rows = clamp_row_est(pathnode->path.rows/rel->ftEntry->segment_number);
```
[the remark this answers is not in the document]
Comment (huiliang-liu) [reply src/backend/optimizer/util/pathnode.c:3071]: you are right. The difference between `all segments` and `master`/`any` can be indicated by locus,  so should not divide segment number here

File: src/backend/optimizer/util/plancat.c [hunk 10]
```diff
@@ -412,10 +412,15 @@ get_relation_info(PlannerInfo *root, Oid relationObjectId, bool inhparent,
...
 	/* Grab the fdwroutine info using the relcache, while we have it */
 	if (relation->rd_rel->relkind == RELKIND_FOREIGN_TABLE)
+	{
 		rel->fdwroutine = GetFdwRoutineForRelation(relation, true);
+		rel->ftEntry->exec_location = GetForeignTable(RelationGetRelid(relation))->exec_location;
+		rel->ftEntry->segment_number = GetForeignTable(RelationGetRelid(relation))->segment_number;
+	}
 	else
+	{
 		rel->fdwroutine = NULL;
```
Comment (adam8157) [inline src/backend/optimizer/util/plancat.c:422]: Shall we give exec_location and ~segment_number~ num_segments default values?

File: src/include/foreign/foreign.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (adam8157) [reply src/include/foreign/foreign.h:72]: `mpp_execute` and ~segment_number~ `num_segments` are general options used by Greenplum kernel, FDW extensions report errors when they see unknown options.

So we extract the info and save it into the struct, for 1, the kernel doesn't need to parse the list every time, 2 separate the info out of list options to not error out.

File: src/backend/optimizer/util/plancat.c (hunk 10 above)
Comment (huiliang-liu) [reply src/backend/optimizer/util/plancat.c:422]: the `num_segments` has default value as the segment count of local server. and the `exec_location ` 's default value is `FTEXECLOCATION_MASTER`

File: src/test/regress/sql/gp_foreign_data.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (Aegeaner) [reply src/test/regress/sql/gp_foreign_data.sql:30]: How will these segments map to foreign server in this case?

File: src/backend/cdb/motion/cdbmotion.c [hunk 11]
```diff
@@ -722,7 +722,7 @@ EndMotionLayerNode(MotionLayerState *mlStates, int16 motNodeID, bool flushCommLa
 	 */
 	if (pMNEntry->preserve_order && pMNEntry->ready_tuple_lists != NULL)
 	{
-		for (i = 0; i < getgpsegmentCount(); i++)
+		for (i = 0; i < pMNEntry->num_senders; i++)
```
Comment (jimmyyih) [inline src/backend/cdb/motion/cdbmotion.c:725]: Do these changes of replacing `getgpsegmentCount()` negatively affect query direct dispatch (e.g. `SELECT * FROM table WHERE a = 1`)?  Presumably, the `num_senders` would be 1 (from `UpdateMotionExpectedReceivers()`) instead of the total segment count with these changes... but I'm not sure if that would cause any issues or not.

Comment (adam8157) [reply src/backend/cdb/motion/cdbmotion.c:725]: Thanks, I missed that point.

Fortunately, this is only for the sorted motion, direct dispatch should be safe.

File: src/include/foreign/foreign.h [hunk 12]
```diff
@@ -53,6 +53,7 @@ typedef struct ForeignServer
 	char	   *serverversion;	/* server version, optional */
 	List	   *options;		/* srvoptions as DefElem list */
 	char		exec_location;  /* execute on MASTER, ANY or ALL SEGMENTS, Greenplum MPP specific */
+	int32		num_segments;	/* the number of segments of the foreign cluster */
```
Comment (denalex) [inline src/include/foreign/foreign.h:56]: what would be a meaning of this for foreign servers that are not Greenplum clusters ? What are the implications of defining this property for foreign servers ? If this only applicable to GP2GP FDW, maybe have this as an option in the list of options so that GP2GP concepts do not leak into a generic framework if it is not using this param ?

File: src/include/foreign/foreign.h [hunk 13]
```diff
@@ -68,10 +69,12 @@ typedef struct ForeignTable
 	Oid			serverid;		/* server Oid */
 	List	   *options;		/* ftoptions as DefElem list */
 	char		exec_location;  /* execute on MASTER, ANY or ALL SEGMENTS, Greenplum MPP specific */
+	int32		num_segments;	/* the number of segments of the foreign cluster */
```
Comment (denalex) [inline src/include/foreign/foreign.h:72]: why would different users be able to specify different values for the number of segments of the foreign server ? Also, if this is really needed, can it go into the list of options instead of being a top-level struct member ?

File: src/test/regress/expected/gp_foreign_data.out [hunk 14]
```diff
@@ -26,3 +26,16 @@ CREATE FOREIGN TABLE ft3 (
...
+CREATE FOREIGN TABLE ft5 (
+       c1 int
+) SERVER s1 OPTIONS (delimiter ',', mpp_execute 'all segments', num_segments '5');
+\d+ ft5
+                            Foreign table "public.ft5"
+ Column |  Type   | Modifiers | FDW Options | Storage | Stats target | Description 
+--------+---------+-----------+-------------+---------+--------------+-------------
+ c1     | integer |           |             | plain   |              | 
+Server: s1
+FDW Options: (delimiter ',', mpp_execute 'all segments', num_segments '5')
```
Comment (denalex) [inline src/test/regress/expected/gp_foreign_data.out:40]: why would we let a table to overwrite the server-level `num_segments` option ?

File: src/include/foreign/foreign.h (hunk 12 above)
Comment (adam8157) [reply src/include/foreign/foreign.h:56]: `exec_location` and `num_segments` are Greenplum APIs for FDWs extending from PG to GP, and they are used by the Greenplum kernel but not the FDW extensions, that's why they are in the generic framework.

File: src/test/regress/expected/gp_foreign_data.out (hunk 14 above)
Comment (adam8157) [reply src/test/regress/expected/gp_foreign_data.out:40]: Greenplum has some corner cases that different tables could have different `num_segments`, for instance, the cluster is in the middle of expanding.

Comment (adam8157) [reply src/test/regress/expected/gp_foreign_data.out]: Made the decision that only supports num_segments option for foreign tables, and updated.

**Review by adam8157 — APPROVED** [review]

**Comment by adam8157** [comment]
Got the PM approval, merging
